### ISSUE TITLE
Cannot specify negative consumer priority

### ISSUE THREAD

**Issue by ralbertazzi** [gmr/pamqp#45]
According to the [RabbitMQ documentation](https://www.rabbitmq.com/consumer-priority.html), consumer priority supports negative numbers too:
> Larger numbers indicate higher priority, and both positive and negative numbers can be used.

However, the following code breaks on `pamqp`:
```python
from pamqp import commands
from pamqp.frame import marshal

frame = commands.Basic.Consume(
    queue="queue",
    no_ack=True,
    exclusive=False,
    consumer_tag="xyz",
    arguments={"x-priority": -1}
)

marshal(frame, 0)
# encode.py", line 195
# struct.error: ubyte format requires 0 <= number <= 255
```

Is this a bug of the library or a RabbitMQ extension of the AMQP specification that is not supported by design?

Thanks!

**Comment by ralbertazzi** [comment]
Negative numbers do seem supported. This is the code that is hit when trying to marshal `-1`:
https://github.com/gmr/pamqp/blob/8e72f7b258789c879c3a80226c4d349bbb3a10e3/pamqp/encode.py#L313-L314

However that `octet` function does not seem to like `-1` 🤷

**Comment by michaelklishin** [comment]
Negative numbers are only used by queue federation. There is no reason for you to use them in your applications.

**Comment by ralbertazzi** [comment]
As the default is 0, I have been using negative priorities in the past to deploy extra less-proritized consumers without replacing existing ones.

Nevertheless, thanks for your input! I'll check that out more in detail. Still, the bug exists nevertheless.

**Comment by gmr** [comment]
> However that octet function does not seem to like -1 🤷

I don't have time to find a better way to handle that right now but will keep this open to circle back in a few months when I'm able to give it attention.

**Comment by saintech** [comment]
-129 works fine, but -128 raise the error

**Comment by michaelklishin** [comment]
No one should need a consumer priority of -128. No one (other than the federation plugin by design) generally should need negative priorities if you have control over what priorities are used by various apps (federation does not, hence it uses a negative value, IIRC no more creative than -1)

**Comment by eandersson** [comment]
> No one should need a consumer priority of -128. No one (other than the federation plugin by design) generally should need negative priorities if you have control over what priorities are used by various apps (federation does not, hence it uses a negative value, IIRC no more creative than -1)

I think what was being said here is that the values `-1` AND `-128` does not work, but `-129` works since values under `-128` are handled by different logic.

**Comment by patstrom** [comment]
This also causes issues with the `x-delivery-limit` parameter, where -1 is a valid value (and even recommended in the [documentation](https://www.rabbitmq.com/docs/quorum-queues#repeated-requeues)).

We want to be able to set this limit explicitly through aiopika (which uses pamqp) just like we did with "normal" pika, but this breaks our queue declarations.

**Comment by michaelklishin** [comment]
@patstrom just to make it clear: the docs mention the value of `-1` as supported. They do not recommend restoring the old behavior of unlimited redeliveries, and the default is 20 for a reason.

**Comment by patstrom** [comment]
Yes that's why I put recommended in quotes. I've happened to come upon an instance where I need to set this value to -1 explicitly since otherwise RabbitMQ returns an error on queue declaration saying I cannot set it to none.
